### PR TITLE
Update elasticsearch

### PR DESCRIPTION
**Pull request by sonalkr132**
[ES 5 requires Java version 1.8.0_73 or later](https://www.elastic.co/guide/en/elasticsearch/reference/current/setup.html#jvm-version)

### CODE REVIEW

**Comment by segiddins** [comment]
We should also update the development instructions

File: app/models/concerns/rubygem_searchable.rb [hunk 1]
```diff
@@ -69,12 +69,12 @@ def self.elastic_search(q) # rubocop:disable Metrics/MethodLength
...
             query do
-              filtered do
+              bool do
                 # Main query, search in name, summary, description
-                query do
+                must do
                   query_string do
                     query q
-                    fields ['name^3', 'summary^1', 'description']
+                    fields ['name^5', 'summary^3', 'description']
```
Comment (dwradcliffe) [inline app/models/concerns/rubygem_searchable.rb]: Any specific reason for this change?

File: test/unit/rubygem_searchable_test.rb [hunk 2]
```diff
@@ -62,8 +62,8 @@ class RubygemSearchableTest < ActiveSupport::TestCase
       example_gem2 = create(:rubygem, name: "example_gem2", downloads: 1)
       example_gem3 = create(:rubygem, name: "example_gem3", downloads: 1)
       create(:version, rubygem: example_gem1)
-      create(:version, rubygem: example_gem2, description: 'some text and keyword')
-      create(:version, rubygem: example_gem3, summary: 'some keyword')
```
Comment (dwradcliffe) [inline test/unit/rubygem_searchable_test.rb]: Why this change? Are all fields required now or something?

File: test/unit/rubygem_searchable_test.rb [hunk 3]
```diff
@@ -41,8 +41,8 @@ class RubygemSearchableTest < ActiveSupport::TestCase
 
   context 'filter' do
     setup do
-      example_1 = create(:rubygem, name: "example_1")
-      example_2 = create(:rubygem, name: "example_2")
```
Comment (dwradcliffe) [inline test/unit/rubygem_searchable_test.rb]: why this change?

File: test/unit/rubygem_searchable_test.rb (hunk 2 above)
Comment (sonalkr132) [inline test/unit/rubygem_searchable_test.rb]: It was a naive attempt to ensure that `avgFieldLength` and `fieldLength` don't affect the score and we test just the boost used on fields (name has the maxmium boost ^5, followed by summary ^3 ). Change was needed because default(in tests) for `summary` is `nil` and that for `description` is `Some awesome gem`.
I have changed it to use same length for all the three fields. [Here](https://gist.github.com/sonalkr132/78bd0ce8c2dcdca5f1e0ac3e21f7efc0) is explain of the current implementation without boost.

File: test/unit/rubygem_searchable_test.rb (hunk 3 above)
Comment (sonalkr132) [inline test/unit/rubygem_searchable_test.rb]: Something fishy is going on here and I don't know how can proceed further. After update, this test was failing because it didn't return any result for query: `example`. I checked the rubygem analyzer:
```
curl -XGET 'localhost:9200/rubygems-development/_analyze?analyzer=rubygem&pretty=true' -d 'example_1'
{
  "tokens" : [
    {
      "token" : "example",
      "start_offset" : 0,
      "end_offset" : 7,
      "type" : "word",
      "position" : 0
    },
    {
      "token" : "1",
      "start_offset" : 8,
      "end_offset" : 9,
      "type" : "word",
      "position" : 1
    }
  ]
}
```
As you can see, it is able to tokenize correctly (I have also added test for rubygems analyzer), however it was not able to find records.

File: app/models/concerns/rubygem_searchable.rb (hunk 1 above)
Comment (sonalkr132) [inline app/models/concerns/rubygem_searchable.rb]: Boost of `^1` is the default. A *keyword* in the summary should be more important than that in the description.

File: test/unit/rubygem_searchable_test.rb (hunk 3 above)
Comment (sonalkr132) [inline test/unit/rubygem_searchable_test.rb]: Changing the name was the easiest workaround, as the test is for `filter` over yanked gems and name of gem doesn't matter here.

**Comment by sonalkr132** [comment]
With help of @karmi (See https://github.com/elastic/elasticsearch-rails/issues/697) I got around the issue with analyzer which had required following (naive)change :anger: :
```diff
-      example_1 = create(:rubygem, name: "example_1")
-      example_2 = create(:rubygem, name: "example_2")
+      example_1 = create(:rubygem, name: "example-1")
+      example_2 = create(:rubygem, name: "example-2")
```
[The failing build](https://travis-ci.org/rubygems/rubygems.org/jobs/221436383#L1038) on `3962e57` is testimony of analyzer quirkiness.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -95,7 +99,7 @@ gems for your organization..**
   * Or install via [alternate methods](https://www.ruby-lang.org/en/downloads/)
 * Use Rubygems 2.6.10
 * Install bundler: `gem install bundler`
-* Install Elastic Search 1.5.2: <https://www.elastic.co/downloads/past-releases/elasticsearch-1-5-2>
+* Install Elastic Search 5.1.2: <https://www.elastic.co/downloads/past-releases/elasticsearch-5-1-2>
```
Comment (dwradcliffe) [inline CONTRIBUTING.md]: I would suggest the docker command here too

**Comment by sonalkr132** [comment]
updated :+1:
